### ISSUE TITLE
Module designer

### ISSUE THREAD

**Issue by AndrewBelt** [VCVRack/Rack#1058]
I've been thinking of how to make module creation stupid-simple, and it's clear to me that a hypothetical module designer tool should not deal with arranging widgets at all. Inkscape etc is 100x better at arranging than Rack can ever be. It should use certain hidden rectangles on the SVG panel to place points on the panel. Right click on the panel to reveal a menu item "Copy C++ widgets" or something, and it will generate source code that should compile with minimal effort.

Running in development mode (`./Rack -d` or `make run`) will enable this feature. The procedure for making a module will be
- Design a panel SVG file. Place rectangles on a certain layer. Only the centers of rectangles will matter
- Clone https://github.com/VCVRack/Template or copy a .cpp file from your existing plugin source code.
- Add it to your plugin by defining a `extern Model *modelWhatever` in your plugin's header file and `plugin->addModel(modelWhatever);` in the main source file.
- Change the panel filename in the `setPanel()` call in your ModuleWidget constructor.
- Launch Rack.
- Add your module. Right click its panel. Select "Copy C++ widgets".
- Paste the source into the ModuleWidget constructor. Change widget types
- Recompile. Your panel should be complete and functional.

**Comment by rubyglow** [comment]
Makes sense Andrew, use the best tool for the job. The missing piece for me is: How do you tell Rack, what kind of widgets those "rectangles" are, so it can generate the right code?

**Comment by AndrewBelt** [comment]
After it generates the code, just replace the class name in the template `createParamCentered`-like functions.

**Comment by freddyz** [comment]
I think it's a great idea.  Have you considered this as a command-line option (eg: ./Rack --generate-module-code newModulePanel.svg) rather than in the GUI of Rack?  It seems unnecessary to actually launch Rack to accomplish this task.

**Comment by AndrewBelt** [comment]
In a way, my [panel2source.py](#374) script does this already. The reason for loading Rack is to generate panel changes from existing modules and tell the generator the plugin and module slugs.

**Comment by lindenbergresearch** [comment]
My current workflow is to layout all in Affinity and then hide the layer of the components and type in all coordinates from Affinity. Thats of course a bit nasty.

I thought about having some invisible objects (boxes) in the .SVG which are named by a special ID (xml tag) and then could be bound via that 'id' on creating of the widgets via the internal create template.
So you could always layout in Inkscape or Affinity etc.  and the coordinates are read on runtime from the SVGs XML structure... (of course with an fallback or so)

What do you think?

**Comment by AndrewBelt** [comment]
See my script above, that's essentially what I already do, except it generates a .cpp file.

**Comment by lindenbergresearch** [comment]
Yes, I know. Maybe I explained it a bit unclear. 
I focus on the fact, that nothing has to be be generated again, all data is read on the fly. So I just layout the boxes and on start the create template figures out the positions.
Else I have always to regenerate code and merge it..
